cpp/server/handler.h

```diff
@@ -32,21 +33,27 @@ class ThreadPool;
 class HttpHandler {
  public:
   // Does not take ownership of its parameters, which must outlive
-  // this instance. The "frontend" parameter can be NULL, in which
-  // case this server will not accept "add-chain" and "add-pre-chain"
-  // requests.
-  HttpHandler(JsonOutput* json_output, LogLookup<LoggedEntry>* log_lookup,
+  // this instance.
+  HttpHandler(LogLookup<LoggedEntry>* log_lookup,
               const ReadOnlyDatabase<LoggedEntry>* db,
               const ClusterStateController<LoggedEntry>* controller,
-              const CertChecker* cert_checker, Frontend* frontend,
-              Proxy* proxy, ThreadPool* pool, libevent::Base* event_base);
+              ThreadPool* pool, libevent::Base* event_base);
   ~HttpHandler();
 
   void Add(libevent::HttpServer* server);
 
- private:
+  void SetOutput(JsonOutput* json_output);
+  void SetProxy(Proxy* proxy);
+
+ protected:
+  // Implemented by subclasses which want to add their own extra http handlers.
+  virtual void AddHandlers(libevent::HttpServer* server) = 0;
```

Comment:
I'm still fuzzy on why it's not just `HttpHandler::Add` that shouldn't just become virtual?